content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
+
+```c
+int* ptr;
+ptr = (int *)malloc(4 * sizeof(int));
+free(ptr);
+ptr[0] = 4;
+return 0;
+```
+
+The naïve solution would be to add instrumentation after every load or store instruction in the LLVM IR, and compare the addresses to our allocation table. However, we run into a series of complications once we look at the actual IR.
+
+```
+%3 = alloca i32, align 4
+%4 = alloca i32, align 4
+%5 = alloca i8**, align 8
+%6 = alloca i32*, align 8
+store i32 0, i32* %3, align 4
+store i32 %0, i32* %4, align 4
+store i8** %1, i8*** %5, align 8
+%7 = call noalias i8* @malloc(i64 16) #3
+%8 = bitcast i8* %7 to i32*
+store i32* %8, i32** %6, align 8
+%9 = load i32*, i32** %6, align 8
+%10 = bitcast i32* %9 to i8*
+call void @free(i8* %10) #3
+%11 = load i32*, i32** %6, align 8
+%12 = getelementptr inbounds i32, i32* %11, i64 0
+store i32 4, i32* %12, align 4
+ret i32 0
+```
+
+First, the program accesses more memory than what was allocated by the programmer through calls to `malloc`. If MemPass compares the address accessed by one of the first `store` instructions with the allocation hash, it wouldn't find the address and emit a false-positive warning.
+
+In reality, programs also allocates stack frame memory with `alloca` instructions. In order to handle these extra allocations, MemPass adds extra instrumentation here, and adds the stack frame addresses/sizes onto the allocation table.
+
+However, there's another problem. Some of the load and store instructions use pointers of arbitrary types. If MemPass doesn't know what the pointer types are, it can't pass those addresses to the runtime library for evaluation.
+
+A solution that MemPass employs is to insert `bitcast` instructions after every load or store instruction, converting the address pointer from its arbitrary type to an 8-bit integer pointer. Since we're only comparing addresses and not the actual values at these addresses, this should work somewhat well. However, the breadth of types in LLVM could allow for some imprecision here.
+
+With `i8*` pointers, all MemPass needs is the size of the memory chunk that a load or store plans to interact with. While this data is not immediately accessible, LLVM provides a handy DataLayout class. After grabbing the type of the element that the *original* pointer points to, MemPass can extract its size and pass that to our library functions.
+
+Finally, we need to actually check if we are accessing memory that is available as per the allocation table. MemPass takes the difference of the address in question with every pointer address in the allocation table, and compares it to the appropriate sizes. If the address is within the bounds of a chunk of memory, then we are fine. Otherwise, the program will emit a use after free warning, and continue to look for more vulnerabilities.
+
+> **Note**: These pointer address comparisons are somewhat sketchy, there likely may be a better approach to memory access checking that we were unable to find.

```

Comment:
TBH I don't see this as sketchy. If your table contains (base address, size) pairs, then it does seem like the natural thing to do with a given access is to look at each pair and see whether the new address is within range for that allocation. Not sure what else you could do!